src/ingest-contact-loaders/csv-upload/index.js

```diff
@@ -0,0 +1,136 @@
+  return;
+}
+
+export function clientChoiceDataCacheKey(organization, campaign, user) {
+  /// returns a string to cache getClientChoiceData -- include items that relate to cacheability
+  //return `${organization.id}-${campaign.id}`;
+  return "";
+}
+
+export async function getClientChoiceData(organization, campaign, user, loaders) {
+  /// data to be sent to the admin client to present options to the component or similar
+  /// The react-component will be sent this data as a property
+  /// return a json object which will be cached for expiresSeconds long
+  /// `data` should be a single string -- it can be JSON which you can parse in the client component
+  return {
+    data: "",
+    expiresSeconds: 0
+  };
+}
+
+export async function processContactLoad(job, maxContacts) {
+  /// trigger processing -- this will likely be the most important part
+  /// you should load contacts into the contact table with the job.campaign_id
+  /// Since this might just *begin* the processing and other work might
+  /// need to be completed asynchronously after this is completed (e.g. to distribute loads)
+  /// AFTER true contact-load completion, this (or another function) MUST call
+  /// src/workers/jobs.js::completeContactLoad(job)
+  ///   The async function completeContactLoad(job) will delete opt-outs, delete duplicate cells, clear/update caching, etc.
+  /// Basic responsibilities:
+  /// 1. delete previous campaign contacts on a previous choice/upload
+  /// 2. set campaign_contact.campaign_id = job.campaign_id on all uploaded contacts
+  /// 3. set campaign_contact.message_status = "needsMessage" on all uploaded contacts
+  /// 4. Ensure that campaign_contact.cell is in the standard phone format "+15551234567"
+  ///    -- do NOT trust your backend to ensure this
+  /// Things to consider in your implementation:
+  /// * Batching
+  /// * Error handling
+  /// * "Request of Doom" scenarios -- queries or jobs too big to complete
+  const campaignId = job.campaign_id;
+  const jobMessages = [];
+
+  await r
+    .knex("campaign_contact")
+    .where("campaign_id", campaignId)
+    .delete();
+
+  // TODO
+  let contacts;
+  if (job.payload[0] === '{') {
+    contacts = JSON.parse(job.payload).contacts;
+  } else {
+    contacts = (await unzipPayload(job)).contacts;
+  }
+  if (maxContacts) {
+    // note: maxContacts == 0 means no maximum
+    contacts = contacts.slice(0, maxContacts);
+  }
+  const chunkSize = 1000;
+
+  const numChunks = Math.ceil(contacts.length / chunkSize);
+
+  for (let index = 0; index < contacts.length; index++) {
+    const datum = contacts[index];
+    if (datum.zip) {
+      // using memoization and large ranges of homogenous zips
+      datum.timezone_offset = await getTimezoneByZip(datum.zip);
+    }
+    datum.campaign_id = campaignId;
+  }
+  for (let index = 0; index < numChunks; index++) {
+    /*
+    await updateJob(job, Math.round((maxPercentage / numChunks) * index))
```

Comment:
why is this commented out? I know I always say to cut commented code, but I really do think that it's a lot easier to read if there isn't commented code left in. If we really need it, can we explain why it's there with another comment?

Reply:
updateJob was necessary during a slow, but large contact upload.  When we deploy on AWS Lambda, there's more a hard-maximum related to data-size and then the max-size is processable pretty quickly.

I have an idea for better statuses in ingest-contact-loaders, if we come back to this work, but even then, it would not rely on the same structure -- statuses reported through JobRequests is inherently flawed, because it's all deleted as soon as the job is complete.